src/runtime/hashmap.h

```diff
@@ -0,0 +1,405 @@
+#ifndef HALIDE_RUNTIME_HASHMAP_H
+#define HALIDE_RUNTIME_HASHMAP_H
+
+// the bulk of this hashmap implementation is based on 'cache.cpp'
+
+#include "printer.h"
+#include "scoped_mutex_lock.h"
+
+namespace Halide { namespace Runtime { namespace Internal {
+
+WEAK bool keys_equal(const uint8_t *key1, const uint8_t *key2, size_t key_size) {
+    return memcmp(key1, key2, key_size) == 0;
+}
+
+WEAK uint32_t djb_hash(const uint8_t *key, size_t key_size) {
+    uint32_t h = 5381;
+    for (size_t i = 0; i < key_size; i++) {
+      h = (h << 5) + h + key[i];
+    }
+    return h;
+}
+
+typedef void(*copy_value_func)(uint8_t *dst, const uint8_t *src, size_t size);
+typedef void(*destroy_value_func)(uint8_t *value, size_t size);
+
+struct CacheEntry {
+    CacheEntry *next;
+    CacheEntry *more_recent;
+    CacheEntry *less_recent;
+    uint8_t *metadata_storage;
+    size_t key_size;
+    uint8_t *key;
+    uint32_t hash;
+    uint32_t in_use_count; // 0 if none returned from halide_cache_lookup
+
+    // The actual stored data.
+    size_t   value_size;
+    uint8_t* value;
+
+    bool init(const uint8_t *cache_key, size_t cache_key_size,
+              uint32_t key_hash,
+              const uint8_t* cache_value, size_t cache_value_size,
+              copy_value_func copy_value);
+    void destroy(destroy_value_func destroy_value);
+};
+
+WEAK bool CacheEntry::init(const uint8_t *cache_key, size_t cache_key_size,
+                           uint32_t key_hash,
+                           const uint8_t* cache_value, size_t cache_value_size,
+                           copy_value_func copy_value) {
+    next = NULL;
+    more_recent = NULL;
+    less_recent = NULL;
+    key_size = cache_key_size;
+    hash = key_hash;
+    in_use_count = 0;
+
+    // Allocate all the necessary space (or die)
+    size_t storage_bytes = 0;
+
+    // First storage for value:
+    storage_bytes += cache_value_size;
+
+    // NOTE(marcos): maybe enforce some alignment between value and key?
+    //storage_bytes = (storage_bytes + 7) / 8;
```

Comment:
We should either enable or remove this.

Reply:
@zvookin I think `cache.cpp` does not explicitly enforce alignment between payload and key here because it relies on the fact that `sizeof(halide_buffer_t) * tuple_count` will result in a byte size that guarantees alignment for the key that follows it. Confirm?

Reply:
In the original code, the result of malloc first stores an array of halide buffers, then an array of dimension structs, and finally the key. This guarantees the required alignment for each of those things. I don't think there is any requirement to align the key. Is the idea that this would somehow improve performance? I suppose we could store the key aligned and padded to make vectorized key comparison faster.